crm_exception/models/crm_lead.py

```diff
@@ -0,0 +1,36 @@
+# Copyright 2023 Quartile Limited
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl).
+
+from odoo import api, models
+from odoo.osv import expression
+
+
+class CrmLead(models.Model):
+    _inherit = ["crm.lead", "base.exception"]
+    _name = "crm.lead"
+    _order = "main_exception_id asc, name desc"
+
+    @api.model
+    def _reverse_field(self):
+        return "crm_lead_ids"
+
+    def _rule_domain(self):
+        base_rule_domain = super()._rule_domain()
+        if self.stage_id:
+            rule_domain = expression.AND(
+                [
+                    base_rule_domain,
+                    [
+                        "|",
+                        ("stage_ids", "in", self.stage_id.ids),
+                        ("stage_ids", "=", False),
+                    ],
+                ]
+            )
+            return rule_domain
+        return base_rule_domain
+
+    @api.constrains("ignore_exception", "stage_id")
+    def _check_quantity_positive(self):
+        for record in self:
+            record._check_exception()
```

Comment:
We should actually extend the write() method to ensure that the check is not restricted to certain fields. For example, with the current logic, even if a field is made required for a specific stage by an exception rule, the field value can be removed once the opportunity is saved with that stage.